stateMutators.js

```diff
@@ -0,0 +1,109 @@
+const gameState = {
+    basketPos: 50,
+    basketValue: 0,
+    score: 0,
+    livesRemaining: 3,
+    gameLevel: 1,
+    gameMode: 'running',
+    fallingObjects: [],
+}
+
+const setBasketPos = (state, value) => ({
+    ...state,
+    basketPos: value,
+});
+
+const moveBasketLeft = (state, value) => ({
+    ...state,
+    basketPos: basketPos - value,
+});
+
+const moveBasketRight = (state, value) => ({
+    ...state,
+    basketPos: basketPos + value,
+});
+
+const moveBasketMouse = (state, value) => ({
+    ...state,
+    basketPos: value,
+});
+
+const setBasketValue = (state, value) => ({
+    ...state,
+    basketValue: value,
+});
+
+const resetBasketValue = (state) => ({
+    ...state,
+    basketValue: 0,
+});
+
+const setScore = (state, value) => ({
+    ...state,
+    score: value,
+});
+
+const resetScore = (state) => ({
+    ...state,
+    score: 0,
+});
+
+const loseLife = (state, value) => ({
+    ...state,
+    livesRemaining: value,
+});
+
+const resetLivesRemaining = (state) => ({
+    ...state,
+    livesRemaining: 3,
+});
+
+const setGameLevel = (state, value) => ({
+    ...state,
+    gameLevel: value,
+});
+
+const resetGameLevel = (state) => ({
+    ...state,
+    gameLevel: 0,
+});
+
+const setGameMode = (state, value) => ({
+    ...state,
+    gameMode: value,
+});
+
+const addFallingObject = (state, fallingObj) => ({
+    ...state,
+    fallingObjects: state.fallingObjects.push(fallingObj)
+});
+
+const removeFallingObject = (state, fallingObjId) => ({
+    ...state,
+    fallingObjects: state.fallingObjects.splice(fallingObjId, 1)
```

Comment:
This `.splice` call has a similar pitfall as the `.push` call above, as it returns the slice of the array that you specify via the two indices you pass in rather than the new array with those values removed. 

e.g., `[1,2,3,4].splice(2,1)` actually returns `[3]`, but you probably want the new state to be `[1,2,4]`.

In a similar way to above, two corresponding ways you could do this:

1. using array destructing + `.slice` calls to create a new array without that element:
```js
const removeFallingObject = (state, fallingObjId) => ({
    ...state,
    fallingObjects: [
		...state.fallingObjects.slice(0, fallingObjId),
		...state.fallingObjects.slice(fallingObjId + 1),
  	],
})
```

2. using `splice` in two lines:
```js
const removeFallingObject = (state, fallingObjId) => {
	state.fallingObjects.splice(fallingObjId, 1)
	return state
}
```

Again, I personally would chose 1 for the reasons above (creates brand new object, more readable), and it's more common in modern JS in my experience

Reply:
roger that